library/core/src/num/mod.rs

```diff
@@ -846,37 +866,32 @@ fn from_str_radix<T: FromStrRadixHelper>(src: &str, radix: u32) -> Result<T, Par
     };
 
     let mut result = T::from_u32(0);
-    if is_positive {
-        // The number is positive
-        for &c in digits {
-            let x = match (c as char).to_digit(radix) {
-                Some(x) => x,
-                None => return Err(PIE { kind: InvalidDigit }),
-            };
-            result = match result.checked_mul(radix) {
-                Some(result) => result,
-                None => return Err(PIE { kind: PosOverflow }),
-            };
-            result = match result.checked_add(x) {
-                Some(result) => result,
-                None => return Err(PIE { kind: PosOverflow }),
-            };
+
+    if radix <= 16 && digits.len() <= mem::size_of::<T>() * 2 - is_signed_ty as usize {
```

Comment:
Can you add a comment why is this done besides just explaining about the safety below? Like we have a fast path that only covers a limited amount of digit for roughly N performance improvements?

Reply:
You could also mention that ideally we'd take the fast path whenever `radix.pow(digits.len()) - 1 <= T::MAX`, but that bound is too slow to compute, so we use this rougher bound and focus on the common case.

Reply:
Thanks Lingman, have added that to the comments.